### ISSUE TITLE
Add typst to crates.io

### ISSUE THREAD

**Issue by twitchyliquid64** [typst/typst#2450]
### Description

Heya, I love Typst and would like to be able to use it from within other Rust code!

I was wondering what the thinking is around filling in the `typst` crates (which are currently parked, presumably by yous) and exposing an API that consumes typst markup and returns either:

 - A PNG or bitmap
 - A mesh or vector path

So it can be integrated with other applications like IDEs?

### Use Case

 - IDEs: An IDE with typst support can render the document in its preview tab.
 - Native integrations: Rust GUI code can render typst markup.

**Comment by tingerrr** [comment]
You can already do so in binary crates using git dependencies!
Adding `typst-library = { git = "https://github.com/typst/typst" }` to your dependencies section. Checkout [typst-bot](https://github.com/mattfbacon/typst-bot) (the bot used on the community discord to render typst from commands) or [typstudio](https://github.com/Cubxity/typstudio) (a standalone typst client).

**Comment by laurmaedje** [comment]
There are multiple reasons I'm still shying away from releasing to crates.io:

- The CLI embeds fonts which are too large for crates.io's 10MB limit. Could maybe be solved through compressed embedding, but that also adds runtime cost.
- Typst releases published to crates.io couldn't have git dependencies. We don't currently have any, but what if we have an issue with an upstream crate? We'd either need to wait a potentially long time until the upstream maintainer makes a release until we merge the fix into Typst or merge early with a temporary git dependency and then our release schedule would be blocked by a third party maintainer. The latter is obviously problematic. Right now we could just make a fork and point to that fork.

**Comment by Kreijstal** [comment]
yeah, it'd be nice to have a pure rust alternative to matplotlib https://github.com/plotters-rs/plotters
this would neeed the 'latex-like' text for equations, but why embed latex when you can use typst instead owo

**Comment by PgBiel** [comment]
> * The CLI embeds fonts which are too large for crates.io's 10MB limit. Could maybe be solved through compressed embedding, but that also adds runtime cost.

Perhaps the Typst compiler (as a library) could be released before the CLI then? I believe the library would be more useful to the Rust ecosystem as a whole.
 
> * Typst releases published to crates.io couldn't have git dependencies. We don't currently have any, but what if we have an issue with an upstream crate? We'd either need to wait a potentially long time until the upstream maintainer makes a release until we merge the fix into Typst or merge early with a temporary git dependency and then our release schedule would be blocked by a third party maintainer. The latter is obviously problematic. Right now we could just make a fork and point to that fork.

I think that, in that case, we can just refrain from publishing further versions until that patch isn't necessary anymore. I know this would "suck" in a way, but we would at least have _some_ versions on crates.io; as is, we are (technically) contributing to this problem by not allowing crates on crates.io to depend on Typst.

**Comment by laurmaedje** [comment]
So you mean making a release but not publishing that to crates.io? Hmm...

**Comment by alerque** [comment]
Instead of forking and using a Git dependency, you can also depend on the released version and include a patch is your app to be applied to it. Especially considering you don't currently have any current cases of this anyway this seems like a strange reason to not want to publish the crate.

**Comment by laurmaedje** [comment]
The Cargo.toml `patches` section is ignored for libraries though. We could patch it for the web app and CLI, but not for a published crate as far as I know.

While tere isn't currently such a case, I suspect that it might happen in the future. For instance, we needed new features in ttf-parser multiple times in the past. Luckily @RazrFalcon is very quick to merge PRs and always makes quick releases. But we might not be so lucky in the future or with another dependency.

I'm not saying that this is a strong enough point to not make crates.io releases, just giving my reasons for not doing it so far.

**Comment by laurmaedje** [comment]
> The main issues isn't being able to get a binary from them anyway, it's the access to the library people want it published for, and for that they can add the font embedding feature if they need it.

The font embedding happens in the binary anyway. Deciding whether we publish CLI and library independently seems sensible.

**Comment by PgBiel** [comment]
> So you mean, if we have a git dependency, still make a release but not publish that to crates.io? Hmm...

Yeah precisely. Basically, we publish the crates we can, until we can't. But the versions published would still be useful.

> The Cargo.toml patches section is ignored for libraries though.

that sucks tbh ;-;

**Comment by twitchyliquid64** [comment]
Concretely, wdyt about publishing to crates.io `typst` and `typst-library`, omitting the font files? Under this model:

 - Integrators would need to bring their own fonts, time source etc (by implementing `typst::World`, like how the [discord bot](https://github.com/mattfbacon/typst-bot/blob/main/worker/src/sandbox.rs) does it)
 - Integrators can [parse typst markup](https://github.com/mattfbacon/typst-bot/blob/main/worker/src/main.rs#L61) with `typst::syntax::parse()`
 - Integrators can [raster](https://github.com/mattfbacon/typst-bot/blob/main/worker/src/render.rs#L66) using `typst::export::render`

We could publish as `0.0.X` to represent the dynamic and changing nature of the API (Semver stability requirements only apply after 1.X.X: https://semver.org/#spec-item-4)

Thoughts?

**Comment by PgBiel** [comment]
> Concretely, wdyt about publishing to crates.io `typst` and `typst-library`, omitting the font files?

I believe only `typst-cli` embeds fonts by default, as a feature. `typst` and `typst-library` don't. So it should be OK as is.

> We could publish as `0.0.X`

Regarding this, I wonder if we could publish older versions retroactively as well (when possible) - that is, publish Typst versions older than 0.8.0 to crates.io too (if we publish something to begin with).

**Comment by laurmaedje** [comment]
Related: https://github.com/typst/typst/issues/108. But it makes sense to keep both issues since we might only publish the library, not the CLI.

**Comment by twitchyliquid64** [comment]
Does rust let you publish a crate that has a path dependency and just put it in at publish time, or does it error out?

If the latter, i think we will need to publish the `typst-macro` and `typst-syntax` crates too. But overall I dont think this will be too hard, just a bit of fiddling to get everything referencing each other to publish properly.

**Comment by Kreijstal** [comment]
you need to publish all dependencies unless they are build dependencies

**Comment by tgross35** [comment]
Regarding the crates.io limit, this can be raised on a per-project basis. You just need to request on Zulip.

(that being said, why not just compress fonts individually for the binary and extract any font the first time it is used? Or this could be a build.rs step.)

**Comment by praseodym** [comment]
v0.11 is available on crates.io: https://crates.io/crates/typst 🥳

**Comment by Kreijstal** [comment]
> v0.11 is available on crates.io: https://crates.io/crates/typst 🥳

better late than never hah

**Comment by twitchyliquid64** [comment]
Thankyou!!! <3 I managed to get this working as a block in my WIP notes app:

![image](https://github.com/typst/typst/assets/6328589/fab39e8d-cfd0-47c3-95b6-81a88e317c1c)

<details>

<summary>The code, open to suggestions</summary>

```rust
mod typst {
    use comemo::Prehashed;
    use once_cell::sync::Lazy;
    use typst::foundations::Bytes;
    use typst::text::{Font, FontBook};
    use typst::Library;

    use typst::diag::FileResult;
    use typst::foundations::Datetime;
    use typst::syntax::{FileId, Source, VirtualPath};
    use typst::World;

    fn library(color: &[u8; 4]) -> &'static Prehashed<Library> {
        use typst::foundations::Smart;
        use typst::layout::{Abs, Margin, PageElem};
        use typst::text::TextElem;

        let mut lib = Library::default();
        lib.styles
            .set(PageElem::set_width(Smart::Custom(Abs::pt(240.0).into())));
        lib.styles.set(PageElem::set_height(Smart::Auto));
        lib.styles
            .set(PageElem::set_margin(Margin::splat(Some(Smart::Custom(
                Abs::pt(3.0).into(),
            )))));

        lib.styles
            .set(TextElem::set_fill(typst::visualize::Paint::Solid(
                typst::visualize::Color::Rgb(typst::visualize::Rgb {
                    color: (
                        color[0] as f32 / 255.0,
                        color[1] as f32 / 255.0,
                        color[2] as f32 / 255.0,
                    )
                        .into(),
                    alpha: 1.0,
                }),
            )));

        // HACK: leaking the library so we can have dynamic colors.
        Box::leak(Box::new(Prehashed::new(lib)))
    }

    static FONTS: Lazy<(Prehashed<FontBook>, Vec<Font>)> = Lazy::new(|| {
        let fonts: Vec<_> = typst_assets::fonts()
            .flat_map(|data| Font::iter(Bytes::from_static(data)))
            .collect();
        let book = FontBook::from_fonts(&fonts);
        (Prehashed::new(book), fonts)
    });

    /// A world for rendering a typst block.
    struct RenderWorld {
        source: Source,
        color: [u8; 4],
    }

    impl World for RenderWorld {
        fn library(&self) -> &Prehashed<Library> {
            &library(&self.color)
        }

        fn book(&self) -> &Prehashed<FontBook> {
            &FONTS.0
        }

        fn main(&self) -> Source {
            self.source.clone()
        }

        fn source(&self, _: FileId) -> FileResult<Source> {
            Ok(self.source.clone())
        }

        fn file(&self, id: FileId) -> FileResult<Bytes> {
            Err(typst::diag::FileError::NotSource)
        }

        fn font(&self, index: usize) -> Option<Font> {
            Some(FONTS.1[index].clone())
        }

        fn today(&self, _: Option<i64>) -> Option<Datetime> {
            Some(Datetime::from_ymd(1970, 1, 1).unwrap())
        }
    }

    pub fn render(code: String, color: [u8; 4]) -> Result<Vec<u8>, String> {
        let id = FileId::new(None, VirtualPath::new("main.typ"));
        let source = Source::new(id, code);
        let world = RenderWorld { source, color };

        use typst::eval::Tracer;
        let mut tracer = Tracer::new();
        let mut document = match typst::compile(&world, &mut tracer) {
            Ok(doc) => doc,
            Err(err) => {
                return Err(err[0].message.clone().into());
            }
        };

        let transparent = typst::visualize::Color::Luma(typst::visualize::Luma {
            color: 0f32.into(),
            alpha: 0.0,
        });
        let pixmap = typst_render::render_merged(
            &document,
            144.0 / 72.0 / 1.6,
            transparent,
            Default::default(),
            transparent,
        );

        Ok(pixmap.encode_png().unwrap())
    }
}
```

</details>

**Comment by laurmaedje** [comment]
@twitchyliquid64 One remark: You should store the `Library` somewhere. The `World::library` function will be called many times by the compiler, not just once.
